src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php

```diff
@@ -82,13 +86,38 @@ public function process(ContainerBuilder $container)
                         $event['method'] = '__invoke';
                     }
                 }
+                $registeredMethods[$event['method']] = true;
 
                 $definition->addMethodCall('addListener', [$event['event'], [new ServiceClosureArgument(new Reference($id)), $event['method']], $priority]);
-
                 if (isset($this->hotPathEvents[$event['event']])) {
                     $container->getDefinition($id)->addTag($this->hotPathTagName);
                 }
             }
+
+            if ($isAutoregisterEnabled) {
+                if (null === ($class = $container->getDefinition($id)->getClass()) || null === ($r = $container->getReflectionClass($class, false))) {
+                    throw new InvalidArgumentException(sprintf('Service "%s" must define the "event" attribute on "%s" tags.', $id, $this->listenerTag));
+                }
+                $hasAutoregisteredMethod = false;
+                foreach ($r->getMethods(\ReflectionMethod::IS_PUBLIC) as $m) {
+                    if ($m->isConstructor() || $m->isDestructor() || $m->isStatic() || $m->isAbstract() || $m->isGenerator() || isset($registeredMethods[$m->getName()])) {
+                        continue;
+                    }
+                    $parameters = $m->getParameters();
+                    if ($m->getNumberOfRequiredParameters() < 1 || !$parameters[0]->hasType() || !\is_a(($eventName = $parameters[0]->getType()->getName()), Event::class, true)) {
```

Comment:
The part about `is_a(Event::class)` I'm not sure about.
Another idea:
- if `__invoke` => always register
- otherwise, the registered methods must match the format used above (`onMyEvent`) - taking only the short class name before the `on`.

BTW, this makes me realize I forgot updating this logic about `onMethods`: declaring `onRequest` should register the method for the short-name part of the event. This would apply for the proposed feature, but also when using an explicit FQCN event. Could you do it in this PR also please?

Reply:
You mean, we don't care about type of event? We just check that method match `onXXX` with `XXX` the shortName of the typehint of the first argument? Event if it's not an Event?

I'm not sure it's a good idea for several reason:
- If a user rename an envet class, the listener won't be called anymore  without triggering error
- User should be free to choose the desired name
- It won't work for a listener that contains several methods pointing to distinguished FQDN but same shortName 

ie for the last point
```
NotificationListener implements EventListenerInterfaca ... {
  public function onEventVersionA(\depsVersionA\Event $event) {
    $this->handle($event->getA());
  }  
  public function onEventVersionB(\depsVersionB\Event $event) {
    $this->handle($event->getFoo()->getA());
  }  
}
```

Reply:
Switch to `__invoke` only for consistency with `MessagerHandlerInterface`.

Fixed `onEvent` for FQDN events